### PR TITLE
Refactoring to suppress eslint warnings

### LINKED ISSUE #3927: Enable the no-param-reassign eslint rule
In #3882, we adopted more eslint rules for our Javascript codebase. The `no-param-reassign` rule is recommended, but it wasn't enabled as part of that PR, since it requires code changes. We should enable the rule and fix the code that fails to lint. We can enable it by removing the override that appears here:

https://github.com/linkerd/linkerd2/blob/0eabbe4173fd8a7da91d5771572bd08d85268275/web/app/.eslintrc#L85-L88

### PR DESCRIPTION
**Pull request by kohsheen1234**
Enabling the eslint/no-param-reassign throws some warnings with the existing code.

Made necessary changes to suppress the warnings

Fixes #3927

Signed-off-by: Kohsheen Tiku <kohsheen.t@gmail.com>

### CODE REVIEW

File: web/app/js/components/MetricsTable.jsx [hunk 1]
```diff
@@ -217,7 +217,8 @@ const preprocessMetrics = metrics => {
 
   _each(tableData, datum => {
     _each(datum.latency, (value, quantile) => {
-      datum[quantile] = value;
+      const datum_ = datum;
+      datum_[quantile] = value;
```
[the remark this answers is not in the document]
Comment (kohsheen1234) [reply web/app/js/components/MetricsTable.jsx]: @cpretzer I am not very sure of this case, but I guess it would work as expected. :)

**Comment by kohsheen1234** [comment]
@Pothulapati I've done my rebase. Please have a look

File: web/app/js/components/MetricsTable.jsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (kohsheen1234) [reply web/app/js/components/MetricsTable.jsx]: @Pothulapati and @cpretzer . There is no need to return 'datum' . I have verified this by interchanging datum by datum_ and the function works as expected.

` const preprocessMetrics = metrics => {
  const tableData = _cloneDeep(metrics);
  _each(tableData, datum_ => {
    _each(datum_.latency, (value, quantile) => {
      const datum = datum_;
      datum[quantile] = value;
    });
  });
  return tableData;
};
`
So, I do not see any need of making further changes. :)

**Comment by kohsheen1234** [comment]
@Pothulapati and @cpretzer . There is no need to return 'datum' . I have verified this by interchanging datum by datum_ and the function works as expected.

` 
const preprocessMetrics = metrics => {
  const tableData = _cloneDeep(metrics);
  _each(tableData, datum_ => {
    _each(datum_.latency, (value, quantile) => {
      const datum = datum_;
      datum[quantile] = value;
    });
  });
  return tableData;
};
`

So, I do not see any need of making further changes. Also I have done my rebase, there are no conflicts any more :)

**Comment by cpretzer** [comment]
@kohsheen1234 thanks this the note! 

I observed the same behavior when I tested the code and was looking to understand why it works that way. I _think_ it has to do with how variables are assigned.

Anyway, everything works, so thanks for the PR! 🙌

File: web/app/.eslintrc [hunk 2]
```diff
@@ -84,6 +84,6 @@
 
     // These rules represent best practices that we're not adhering to, and should be enabled
     "jsx-a11y/click-events-have-key-events": 0,
-    "no-param-reassign": 0
+    "no-param-reassign": 1
```
Comment (alpeb) [inline web/app/.eslintrc]: I think it's better to just delete this line to enable that rule

File: web/app/js/components/ErrorModal.jsx [hunk 3]
```diff
@@ -52,20 +52,21 @@ class ErrorModal extends React.Component {
 
     const byPodAndContainer = _map(errorsByPod, (podErrors, pod) => {
       const byContainer = _reduce(podErrors.errors, (errors, err) => {
+        const errors__ = errors;
```
Comment (alpeb) [inline web/app/js/components/ErrorModal.jsx]: nit: could you use a single underscore, like you did in your other changes?

**Comment by alpeb** [comment]
Actually, I take my previous comment back :blush: 
When doing `const datum_ = datum;` the new var receives a reference, not a copy of `datum`, so mutating it does mutate the original one.
That being said, I believe creating new variables is sort of missing the spirit of the rule. I think it's fine, when the argument is an object, to allow to modify its properties. So I would suggest we relax the rule with `{ "props": true }`. What do you think?

**Comment by cpretzer** [comment]
> Actually, I take my previous comment back 😊
> When doing `const datum_ = datum;` the new var receives a reference, not a copy of `datum`, so mutating it does mutate the original one.
> That being said, I believe creating new variables is sort of missing the spirit of the rule. I think it's fine, when the argument is an object, to allow to modify its properties. So I would suggest we relax the rule with `{ "props": true }`. What do you think?

👍 I think that makes sense and makes for more readable/understandable code.

**Comment by kohsheen1234** [comment]
> Actually, I take my previous comment back 😊
> When doing `const datum_ = datum;` the new var receives a reference, not a copy of `datum`, so mutating it does mutate the original one.
> That being said, I believe creating new variables is sort of missing the spirit of the rule. I think it's fine, when the argument is an object, to allow to modify its properties. So I would suggest we relax the rule with `{ "props": true }`. What do you think?

By default, in the rule eslint/no-param-reassign we have, `{ "props": false}` , which means that we can directly assign values to the properties of the object which is passed as an argument. By making`{"prop":true}`, we will simply make the rule more strict instead of making it relaxed.

params

File: web/app/.eslintrc (hunk 2 above)
Comment (kohsheen1234) [reply web/app/.eslintrc]: @alpeb  I will do the same, thank you for the suggestion :)

File: web/app/js/components/ErrorModal.jsx (hunk 3 above)
Comment (kohsheen1234) [reply web/app/js/components/ErrorModal.jsx]: Surely, I'll use a single underscore. :)

**Comment by alpeb** [comment]
> By default, in the rule eslint/no-param-reassign we have, `{ "props": false}` , which means that we can directly assign values to the properties of the object which is passed as an argument. By making`{"prop":true}`, we will simply make the rule more strict instead of making it relaxed.

You're right. But then I'm puzzled: why was something like `NetworkGraph.jsx` giving warnings if we were just changing object properties?

File: web/app/.eslintrc (hunk 2 above)
Comment (kohsheen1234) [reply web/app/.eslintrc]: By using this line `"no-param-reassign": ["error", { "props": false }] ` , we can reassign the properties of the objects that are passed as parameters.

Comment (alpeb) [reply web/app/.eslintrc]: That looks good :+1:

**Comment by kohsheen1234** [comment]
> > By default, in the rule eslint/no-param-reassign we have, `{ "props": false}` , which means that we can directly assign values to the properties of the object which is passed as an argument. By making`{"prop":true}`, we will simply make the rule more strict instead of making it relaxed.
> 
> You're right. But then I'm puzzled: why was something like `NetworkGraph.jsx` giving warnings if we were just changing object properties?

@alpeb Even I got confused, I don't know why it shows warnings. But I realized that using this line `"no-param-reassign":` ["error", { "props": false }]` does not show those warnings. 

Also, I had a doubt: After making changes by assigning values to properties of the object, I feel the dashboard isn't working as expected, though I am not sure about the same.

This is what I get after making the necessary changes. The first one is showing 0/2.

deploy

**Comment by alpeb** [comment]
Thanks fine :slightly_smiling_face: 
It just means the `coredns` pod has two replicas, none of which has been injected by linkerd, which is expected since none of the pods from the kube-system namespace should be injected.

**Comment by kohsheen1234** [comment]
@alpeb Thank you for the clarification. 👍  I have made changes as requested in the Pull Request. Can you just check if any further changes are required. :)

File: web/app/js/components/MetricsTable.jsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (alpeb) [reply web/app/js/components/MetricsTable.jsx]: After having added the `"no-param-reassign": ["error", { "props": false }]` rule, you no longer need to edit this file, nor `Tap.jsx`  :wink:

File: web/app/js/components/TopModule.jsx [hunk 4]
```diff
@@ -202,37 +202,38 @@ class TopModule extends React.Component {
   }
 
   indexTopResult = (d, topResults) => {
+    const topResults_ = topResults;
```
Comment (alpeb) [inline web/app/js/components/TopModule.jsx]: This change is no longer required either

Comment (kohsheen1234) [reply web/app/js/components/TopModule.jsx]: Yes. You are right. :) Just made the changes.

**Comment by kohsheen1234** [comment]
@alpeb Can you check now. I have made the changes you told. Thank you very much for helping me with all these changes :)

File: web/node_modules/.yarn-integrity [hunk 5]
```diff
@@ -0,0 +1,10 @@
+{
+  "systemParams": "linux-x64-57",
+  "modulesFolders": [],
+  "flags": [],
+  "linkedModules": [],
+  "topLevelPatterns": [],
+  "lockfileEntries": {},
+  "files": [],
+  "artifacts": {}
+}
```
Comment (alpeb) [inline web/node_modules/.yarn-integrity]: Could you please remove this file?

File: web/yarn.lock [hunk 6]
```diff
@@ -0,0 +1,4 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
+# yarn lockfile v1
+
+
```
Comment (alpeb) [inline web/yarn.lock]: Could you also please remove this file?

File: web/app/js/components/util/Utils.js [hunk 7]
```diff
@@ -119,10 +121,11 @@ export const toClassName = name => {
  Create regex string from user input for a filter
 */
 export const regexFilterString = input => {
+  let input_ = input;
   // make input lower case and strip out unwanted characters
-  input = input.replace(/[^A-Z0-9/.\-_*]/gi, '').toLowerCase();
+  input_ = input_.replace(/[^A-Z0-9/.\-_*]/gi, '').toLowerCase();
```
Comment (alpeb) [inline web/app/js/components/util/Utils.js]: I think you can transform these into a single liner:
```javascript
let input_ = input.replace(/[^A-Z0-9/.\-_*]/gi, '').toLowerCase();
```

Comment (kohsheen1234) [reply web/app/js/components/util/Utils.js]: Done :)

File: web/yarn.lock (hunk 6 above)
Comment (kohsheen1234) [reply web/yarn.lock]: Done

File: web/node_modules/.yarn-integrity (hunk 5 above)
Comment (kohsheen1234) [reply web/node_modules/.yarn-integrity]: Done :)

**Comment by kohsheen1234** [comment]
> Thanks @kohsheen1234 !
> I've just got a few more minor comments, and then this should be ready to merge

@alpeb Thank you very much. :) . I made the new changes as well.

**Comment by kohsheen1234** [comment]
> Thanks @kohsheen1234 for addressing the feedback so quickly 
> 
> As a first time contributor we'd love to send you some swag! Do you mind filling out this [form](https://docs.google.com/forms/d/19GLBNfu8XnIETrwEI3ijoWlocAiDp3-j7CW3l4cTnyc/viewform)?

Thank you very much @alpeb for your guidance. I have filled the form for some swag. :+1:

**Pull request merged**